File: ruby/lib/coding_interview/presentation/asset_controller.rb
require_relative "presentation_exception"
require_relative "presentation_preparation"
require_relative "../application/usecase/asset/get_asset_usecase"

module CodingInterview
  module Presentation
    StockDto = Struct.new(:symbol, :evaluation_amount)
    GetAssetRequest = Struct.new(:user_id)
    GetAssetResponse = Struct.new(:cash_amount, :stocks)

    class AssetController
      include PresentationPreparation

      def initialize(get_asset_usecase)
        @get_asset_usecase = get_asset_usecase
      end

      def get_asset(req)
        uid = parse_user_id(req.user_id)
        out =
          begin
            @get_asset_usecase.run(Application::Usecase::Asset::GetAssetUsecaseInput.new(uid))
          rescue Application::Usecase::Asset::UserNotFound
            raise BadRequestException.new("user not found")
          end
        GetAssetResponse.new(
          out.cash_amount.to_s("F"),
          out.stocks.map { |e| StockDto.new(e.symbol.to_s, e.evaluation_amount.to_s("F")) }
        )
      end
    end
  end
end

File: python/src/coding_interview/presentation/asset_controller.py
from __future__ import annotations

from dataclasses import dataclass

from coding_interview.application.usecase.asset.get_asset_usecase import (
    GetAssetUsecase,
    GetAssetUsecaseInput,
)
from coding_interview.application.usecase.exceptions import UserNotFoundError
from coding_interview.presentation.exceptions import BadRequestException
from coding_interview.presentation.preparation import parse_user_id


@dataclass(frozen=True)
class StockDto:
    symbol: str
    amountJpy: str


@dataclass(frozen=True)
class GetAssetRequest:
    userId: str


@dataclass(frozen=True)
class GetAssetResponse:
    cashAmount: str
    stocks: tuple[StockDto, ...]


class AssetController:
    def __init__(self, get_asset_usecase: GetAssetUsecase) -> None:
        self._get_asset_usecase = get_asset_usecase

    def get_asset(self, req: GetAssetRequest) -> GetAssetResponse:
        uid = parse_user_id(req.userId)
        try:
            out = self._get_asset_usecase.run(GetAssetUsecaseInput(uid))
        except UserNotFoundError:
            raise BadRequestException("user not found")
        return GetAssetResponse(
            cashAmount=str(out.cash_amount),
            stocks=tuple(StockDto(symbol=str(s.symbol), amountJpy=str(s.amount_jpy)) for s in out.stocks),
        )
